lib/action_view/component/base.rb

```diff
@@ -115,21 +115,21 @@ def has_initializer?
         end
 
         def source_location
-          # Require #initialize to be defined so that we can use
-          # method#source_location to look up the file name
-          # of the component.
-          #
-          # If we were able to only support Ruby 2.7+,
-          # We could just use Module#const_source_location,
-          # rendering this unnecessary.
-          raise NotImplementedError.new("#{self} must implement #initialize.") unless has_initializer?
-
-          instance_method(:initialize).source_location[0]
+          if RUBY_VERSION < "2.7"
+            # When using a version of Ruby < 2.7, require #initialize
+            # to be defined so that we can use method#source_location
+            # to look up the file name of the component.'
+            raise NotImplementedError.new("#{self} must implement #initialize.") unless has_initializer?
+
+            return instance_method(:initialize).source_location[0]
+          end
+
+          Module.const_source_location(self.name)[0]
         end
 
         def eager_load!
           self.descendants.each do |descendant|
-            descendant.compile if descendant.has_initializer?
+            descendant.compile if RUBY_VERSION >= "2.7" || descendant.has_initializer?
```

Comment:
Seems like this should be tested. But I'm not sure if I want to test `#eager_load!`

Maybe this should be refactored so the logic can be tested separately. Perhaps by replacing `has_initializer?` with a method like `can_find_template?` that makes sense for both strategies of finding the source. I guess this sort of refactor would be good because it would reduce references to the initializer in the API, too.

Reply:
I had the same thought. See #172 where I moved this logic into the Railtie. Leaking these detail of the API feels pretty icky. IMO we should just call `compile` here and let it do the no-op if there is no initializer

Reply:
I'm going to throw up a quick PR to clean this stuff up that I think will make this easier.

Reply:
See #175. On the back of that change you'd just need to change `source_location` for Ruby > 2.7 and wouldn't need to worry about it here.